Project1/reference_logic.h

```diff
@@ -0,0 +1,22 @@
+#pragma once
+
```

Comment:
Why? This is scary, because reference_wrapper is implicitly convertible to reference, so it makes sense to do `a == b` to try to get value equals, but you're going to get pointer equals instead.